### ISSUE TITLE
Bug: RecursionError when handler path is same as static file path

### ISSUE THREAD

**Issue by AgarwalPragy** [litestar-org/litestar#2629]
### Description

I'm trying to setup the following
- `GET /uploads` shows a HTML form for uploading file
- `POST /uploads` for uploading file
- `GET /uploads/[path]` for serving the uploaded file

### MCVE

```python
from typing import Annotated
from uuid import uuid4 as uuid

from litestar import Litestar
from litestar import get, Controller
from litestar import post
from litestar.datastructures import UploadFile
from litestar.enums import MediaType
from litestar.enums import RequestEncodingType
from litestar.params import Body
from litestar.static_files import StaticFilesConfig


class UploadController(Controller):
    path = '/uploads'

    @get(media_type=MediaType.HTML)
    def file_upload_form(self) -> str:
        return f"""
        <html>
            <body>
                <form action="{self.path}" enctype="multipart/form-data" method="post">
                    <input type="file" accept="image/png, image/jpeg" name="file" />
                    <input type="submit" value="Submit">
                </form>
            </body>
        </html>
        """

    @post(media_type=MediaType.TEXT, sync_to_thread=False)
    async def handle_file_upload(
            self,
            data: Annotated[UploadFile, Body(media_type=RequestEncodingType.MULTI_PART)],
    ) -> str:
        contents = await data.read()
        file_id = str(uuid())
        filepath = f"uploads/{file_id}_{data.filename}"
        with open(filepath, "wb") as f:
            f.write(contents)
        return filepath


app = Litestar(
    [UploadController],
    static_files_config=[
        StaticFilesConfig(directories=["uploads"], path="/uploads"),
    ],
)
```

### Logs

```bash
litestar run --debug --reload
/app.py:17: LitestarWarning: Use of a synchronous callable <function UploadController.file_upload_form at 0x104c35260> without setting sync_to_thread is discouraged since synchronous callables can block the main thread if they perform blocking operations. If the callable is guaranteed to be non-blocking, you can set sync_to_thread=False to skip this warning, or set the environmentvariable LITESTAR_WARN_IMPLICIT_SYNC_TO_THREAD=0 to disable warnings of this type entirely.
  @get(media_type=MediaType.HTML)
/app.py:30: LitestarWarning: Use of an asynchronous callable <function UploadController.handle_file_upload at 0x104c353a0> with sync_to_thread; sync_to_thread has no effect on async callable. You can disable this warning by setting LITESTAR_WARN_SYNC_TO_THREAD_WITH_ASYNC=0
  @post(media_type=MediaType.TEXT, sync_to_thread=False)
Traceback (most recent call last):
  File "/venv/bin/litestar", line 8, in <module>
    sys.exit(run_cli())
             ^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/litestar/__main__.py", line 6, in run_cli
    litestar_group()
  File "/venv/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/rich_click/rich_command.py", line 125, in main
    with self.make_context(prog_name, args, **extra) as ctx:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/litestar/cli/_utils.py", line 250, in make_context
    self._prepare(ctx)
  File "/venv/lib/python3.11/site-packages/litestar/cli/_utils.py", line 232, in _prepare
    env = ctx.obj = LitestarEnv.from_env(ctx.params.get("app_path"), ctx.params.get("app_dir"))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/litestar/cli/_utils.py", line 127, in from_env
    loaded_app = _autodiscover_app(cwd)
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/litestar/cli/_utils.py", line 340, in _autodiscover_app
    module = importlib.import_module(import_path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/app.py", line 43, in <module>
    app = Litestar(
          ^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/litestar/app.py", line 456, in __init__
    self.register(static_config.to_static_files_app())
  File "/venv/lib/python3.11/site-packages/litestar/app.py", line 628, in register
    self.asgi_router.construct_routing_trie()
  File "/venv/lib/python3.11/site-packages/litestar/_asgi/asgi_router.py", line 142, in construct_routing_trie
    validate_node(node=self.root_route_map_node)
  File "/venv/lib/python3.11/site-packages/litestar/_asgi/routing_trie/validate.py", line 45, in validate_node
    validate_node(node=child)
  File "/venv/lib/python3.11/site-packages/litestar/_asgi/routing_trie/validate.py", line 45, in validate_node
    validate_node(node=child)
  File "/venv/lib/python3.11/site-packages/litestar/_asgi/routing_trie/validate.py", line 45, in validate_node
    validate_node(node=child)
  [Previous line repeated 972 more times]
  File "/venv/lib/python3.11/site-packages/litestar/_asgi/routing_trie/validate.py", line 33, in validate_node
    and any(
        ^^^^
  File "/venv/lib/python3.11/site-packages/litestar/_asgi/routing_trie/validate.py", line 35, in <genexpr>
    list(child.path_parameters.values())
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RecursionError: maximum recursion depth exceeded while calling a Python object
```

### Litestar Version

2.3.1

### Platform

- [ ] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2629">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2629/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2629/pledge.svg">
</picture>
</a>

**Comment by provinzkraut** [comment]
@AgarwalPragy Thanks for the report, we'll look into it!

This has nothing to do with controllers though and can be reproduced by this very simply application:

```python
from litestar import Litestar
from litestar import get
from litestar.static_files import StaticFilesConfig


@get("/uploads")
async def handler() -> None:
    pass


app = Litestar(
    [handler],
    static_files_config=[
        StaticFilesConfig(directories=["uploads"], path="/uploads"),
    ],
)
```

**Comment by provinzkraut** [comment]
@AgarwalPragy I've put up a PR that should fix this: #2630.

**Comment by AgarwalPragy** [comment]
@provinzkraut this fixes the RecursionError, but doesn't accomplish the thing that was desired. The static file path seems to take precedence over the routes - the `/uploads` path returns the following

```json
{
  "status_code": 404,
  "detail": "no file or directory match the path uploads was found"
}
```

**Comment by provinzkraut** [comment]
@AgarwalPragy Thanks for bringing this up again, I had totally missed this aspect! 

Unfortunately this one doesn't seem to be an easy fix like before so it might take a while to sort it out.

**Comment by provinzkraut** [comment]
@peterschutt once you're back you might want to take a look at this as well. 

The issue seems to be that there's a priority conflict between the regular mounts and the static file app; We want to resolve mounts before all other handlers, but since mounts don't have have any method affinity, the mount for the static file app will always match, even for methods it can't handle.

The only two solutions I could come up with on the spot are:

1. Give mounts the option to specify methods as well (would be another special casing just for the static files app, I don't like this)
2. Remove the special treatment of the static files app completely and instead build it with route handlers. This will probably negatively impact performance.

**Comment by provinzkraut** [comment]
@AgarwalPragy I think this might be more of a feature request than a bug report. 

Basically, our current architecture doesn't allow this since the static files are hosted within a separate ASGI application, which does *not* allow other routes on the same path, since ASGI apps are a black box from the outside, meaning the router has no information about which sub routes or HTTP methods this app might handle.

What *should* happen though is that you get a configuration error during setup for this.

**Comment by AgarwalPragy** [comment]
> @AgarwalPragy I think this might be more of a feature request than a bug report.
> 
> Basically, our current architecture doesn't allow this since the static files are hosted within a separate ASGI application, which does _not_ allow other routes on the same path, since ASGI apps are a black box from the outside, meaning the router has no information about which sub routes or HTTP methods this app might handle.
> 
> What _should_ happen though is that you get a configuration error during setup for this.

That's fine - if the static files route throws a configuration error, the programmer will know that they should
1. either not have routes overlapping with the static files config
2. or they should remove the static files config, and serve files themselves through normal routes.

At present, it seems like a bug since it fails silently.

**Comment by peterschutt** [comment]
> What should happen though is that you get a configuration error during setup for this.

This sounds like the right solution.

**Comment by provinzkraut** [comment]
@AgarwalPragy This took a while, but there's a potential fix in #2960.

I opted to go against my earlier judgement that we should simply raise a configuration error here, mainly for reasons of practicality. After digging through our static files code I found that it's way more complicated than it needs to be, which is what leads to cases like this. You'll also find that the example works fine when everything happens on the *root* path, so e.g. `/`. Having other route handler occupy the same path, with the same method as an ASGI app shouldn't be possible at all, but the static files app adds a lot of special casing to make it possible.

Basically, the short version is: The most reasonable solution here is to make the static file handling *not* rely on special treatment everywhere to make it work, and not to add *more* special casing to fix this particular bug.

**Comment by provinzkraut** [comment]
Fixed in #2960. This will be included in the 2.6.0 release.
